src/category_theory/yoneda.lean

```diff
@@ -127,36 +118,150 @@ If `coyoneda.map f` is an isomorphism, so was `f`.
+  (λ X, { hom := λ f, f ⟨⟩, inv := λ x _, x })
+  (by tidy)
+
+-- /-- A Type-valued presheaf `P` is isomorphic to the composition of `P` with the
+--   coyoneda functor coming from `punit`. -/
+-- @[simps] def iso_comp_punit (P : C ⥤ Type v₁) : (P ⋙ coyoneda.obj (op punit.{v₁+1})) ≅ P :=
+-- iso_whisker_left P punit_iso ≪≫ P.right_unitor
 
 end coyoneda
 
+namespace functor
+
+
 /--
-A presheaf `F` is representable if there is object `X` so `F ≅ yoneda.obj X`.
+A functor `F : Cᵒᵖ ⥤ Type v₁` is representable if there is object `X` so `F ≅ yoneda.obj X`.
 
 See https://stacks.math.columbia.edu/tag/001Q.
 -/
--- TODO should we make this a Prop, merely asserting existence of such an object?
-class representable (F : Cᵒᵖ ⥤ Type v₁) :=
-(X : C)
-(w : yoneda.obj X ≅ F)
+class representable (F : Cᵒᵖ ⥤ Type v₁) : Prop :=
+(has_representation : ∃ X (f : yoneda.obj X ⟶ F), is_iso f)
 
-end category_theory
+/-- Given an object `X` and an isomorphism `yoneda.obj X ≅ F`, `F` is representable. -/
+lemma representable_of_nat_iso {F : Cᵒᵖ ⥤ Type v₁} (X : C) (i : yoneda.obj X ≅ F) :
+  F.representable :=
+{ has_representation := ⟨X, i.hom, infer_instance⟩ }
 
-namespace category_theory
--- For the rest of the file, we are using product categories,
--- so need to restrict to the case morphisms are in 'Type', not 'Sort'.
+instance {X : C} : representable (yoneda.obj X) :=
+representable_of_nat_iso _ (iso.refl _)
+
+/--
+A functor `F : C ⥤ Type v₁` is corepresentable if there is object `X` so `F ≅ coyoneda.obj X`.
+
+See https://stacks.math.columbia.edu/tag/001Q.
+-/
+class corepresentable (F : C ⥤ Type v₁) : Prop :=
+(has_corepresentation : ∃ X (f : coyoneda.obj X ⟶ F), is_iso f)
+
+lemma corepresentable_of_nat_iso {F : C ⥤ Type v₁} (X : Cᵒᵖ) (i : coyoneda.obj X ≅ F) :
+  F.corepresentable :=
+{ has_corepresentation := ⟨X, i.hom, infer_instance⟩ }
+
+instance {X : Cᵒᵖ} : corepresentable (coyoneda.obj X) :=
+corepresentable_of_nat_iso X (iso.refl _)
+
+instance : corepresentable (𝟭 (Type v₁)) :=
+corepresentable_of_nat_iso (op punit) coyoneda.punit_iso
+
+section representable
+variables (F : Cᵒᵖ ⥤ Type v₁)
+variable [F.representable]
+
+/-- The representing object for the representable functor `F`. -/
+noncomputable def repr_X : C :=
+(representable.has_representation : ∃ X (f : _ ⟶ F), _).some
+
+/-- The (forward direction) of the isomorphism witnessing `F` is representable. -/
+noncomputable def repr_f : yoneda.obj F.repr_X ⟶ F :=
+representable.has_representation.some_spec.some
+
+/--
+The representing element for the representable functor `F`, sometimes called the universal
+element of the functor.
+-/
+noncomputable def repr_x : F.obj (op F.repr_X) :=
+F.repr_f.app (op F.repr_X) (𝟙 F.repr_X)
+
+instance : is_iso F.repr_f :=
+representable.has_representation.some_spec.some_spec
+
+/--
+An isomorphism between `F` and a functor of the form `C(-, F.repr X)`.  Note the components
+`F_repr.w.app X` definitionally have type `(X.unop ⟶ F.repr_X) ≅ F.obj X`.
```

Comment:
```suggestion
`F_repr_w.app X` definitionally have type `(X.unop ⟶ F.repr_X) ≅ F.obj X`.
```

Also, why `repr_w` and not just `repr`?

Reply:
Fair point - but I needed some way to distinguish `repr_X`, `repr_f` and `repr_x` - if you have better names feel free to share!

Reply:
On second thought, I don't actually mind the name `repr_w` all that much. Feel free to keep it if you would like. I think `repr_X`, `repr_f` and `repr_x` are all fine as is.